float64[2] a
float64[2] b
float64[2] c
float64[2] d
